### PR TITLE
Replace Visibility.isMoreRestrictiveThan with opCmp

### PR DESCRIPTION
**Pull request by Geod24**

### CODE REVIEW

File: src/dmd/dsymbol.d [hunk 1]
```diff
@@ -138,20 +138,20 @@ struct Prot
...
     }
 
-    extern (C++):
-
     /**
      * Checks if `this` is superset of `other` restrictions.
      * For example, "protected" is more restrictive than "public".
      */
-    bool isMoreRestrictiveThan(const Prot other) const
+    int opCmp(const Prot other) const pure nothrow @nogc @safe
```
Comment (PetarKirov) [inline src/dmd/dsymbol.d:152]: The comment above could better be explained with a ddoc-ed unittest.

**Review by PetarKirov — APPROVED** [review]

**Comment by UplinkCoder** [comment]
@Geod24 I find this change obscures meaning.

**Comment by UplinkCoder** [comment]
is `protected` greater than `private`? 
What does that even mean?
Is `protected` more restrictive than `private`? - no.

Please reconsider.

File: src/dmd/dsymbol.d (hunk 1 above)
Comment (Geod24) [reply src/dmd/dsymbol.d:161]: Good point, extended the documentation and added a documented unittest.

**Comment by Geod24** [comment]
> is `protected` greater than `private`?

`protected` is less restrictive than `private`. `opCmp` is meant to order items within a set, and that's exactly what it's doing here. It's one of the rare case where `opEquals` and `opCmp` might disagree and it's sensible.

**Comment by UplinkCoder** [comment]
@Geod24 but why not leave the function in place which documents it's semantics,
rather than saying 
```
Let {private, protected, public, package} be the the of protections P.
Within P the ordering of the operator '<' lessThan shall be such that package < protected < public < public.
```

**Comment by Geod24** [comment]
Because `isMoreRestrictiveThan` is just a specialization of `opCmp`, as obvious from the diff.
And users tend to more familiar with the later.

**Comment by UplinkCoder** [comment]
> Because `isMoreRestrictiveThan` is just a specialization of `opCmp`, as obvious from the diff.
> And users tend to more familiar with the later.

Which Users?

**Comment by UplinkCoder** [comment]
The function is not used enough that it makes sense to introduce the concept of the protection set ordering. 
into the context of everyone who wants to understand the code.

**Comment by Geod24** [comment]
> Which Users?

`opCmp` is a basic tool in the D programmer's toolbox. `isMoreRestrictiveThan` has cognitive overhead because one has to look up what it is doing. Likewise, ordering between visibility is rather obvious from my experience.

> The function is not used enough that it makes sense to introduce the concept of the protection set ordering.

The concept is already there.

**Comment by UplinkCoder** [comment]
`isMoreRestrictiveThan` does not require lookup.
what it returns is in the name.
`<` does require lookup.

**Comment by UplinkCoder** [comment]
I don't care that much in this case, since it's not used a whole lot.
But I find the direction to be obscuring meaning rather than clarifying.

**Comment by UplinkCoder** [comment]
> ordering between visibility is rather obvious from my experience

Well it can go either way.

you can define the order on restricitive in which lessThan means less restrictive, in which public is less than private, because it's less restrictive. 
or
you can define the order on visibility in which case private is less than public because it's less visible. 

In that way less than by itself is ambiguous

**Comment by Geod24** [comment]
As discussed, the names have been updated from "protection" to "visibility", so it is not ambiguous anymore. Care to merge ?

File: src/dmd/frontend.h [hunk 2]
```diff
@@ -970,7 +970,7 @@ struct Visibility
 {
     Kind kind;
     Package* pkg;
-    bool isMoreRestrictiveThan(const Visibility other) const;
```
Comment (UplinkCoder) [inline src/dmd/frontend.h:973]: How do we access opCmp from c++?

Comment (Geod24) [reply src/dmd/frontend.h:973]: It's not used from C++, as discussed on Slack.

File: src/dmd/dsymbol.d [hunk 3]
```diff
@@ -138,20 +138,37 @@ struct Visibility
...
 
-    extern (C++):
-
     /**
-     * Checks if `this` is superset of `other` restrictions.
-     * For example, "protected" is more restrictive than "public".
+     * Checks if `this` is less or more restrictive than `other`
+     *
+     * Params:
+     *   other = Protection to compare `this` to.
```
Comment (WalterBright) [inline src/dmd/dsymbol.d:152]: s/Protection/Visibility/

File: src/dmd/dsymbol.d [hunk 4]
```diff
@@ -138,20 +138,37 @@ struct Visibility
...
     {
         this.kind = kind;
     }
 
-    extern (C++):
-
     /**
-     * Checks if `this` is superset of `other` restrictions.
-     * For example, "protected" is more restrictive than "public".
+     * Checks if `this` is less or more restrictive than `other`
```
Comment (WalterBright) [inline src/dmd/dsymbol.d:149]: The word "restrictive" has a dubious connection to "visibility".

File: src/dmd/dsymbol.d [hunk 5]
```diff
@@ -138,20 +138,37 @@ struct Visibility
...
+     *
+     * Params:
+     *   other = Protection to compare `this` to.
+     *
+     * Returns:
+     *   A value `< 0` if `this` is less restrictive than `other`,
+     *   a value `> 0` if `this` is more restrictive than `other`,
+     *   and `0` if they are at the same level.
+     *   Note that `package` protection with different packages
```
Comment (WalterBright) [inline src/dmd/dsymbol.d:158]: `protection` again

**Review by WalterBright — DISMISSED** [review]
see my code comments

**Comment by Geod24** [comment]
Comments addressed

**Review by RazvanN7 — APPROVED** [review]

**Review by RazvanN7 — APPROVED** [review]
`opEquals` from `frontend.h` should also be deleted as it is no longer `extern(C++)`. This is the reason why circleCi is failing. Please fix this and let's merge this.

**Pull request merged**